feat: update model type definitions

lihbr opened the pull request:
wip

lihbr commented:
OK, so I refactored the PR following our discussion so that we only focus on having external types (these types) being compatible with internal types.

```typescript
const fn = (obj: ExternalType) => {};

const internal: InternalType = {};
const external: ExternalType = {};

fn(internal); // OK
fn(external); // OK
```

I tried to shave it down to what made sense, however, I couldn't get rid of:
- `Range` type being an alias of the `Number` type (valid but undocumented/not maintained)
- `Separator` type being deprecated and broken (crashes on document saving)
- `LegacySlice` type being deprecated and broken (crashes custom type editor)

Let me know if you think those updated types seem manageable by `prismic-ts-codegen` and if you have any better idea on how to better hide the 3 above~

angeloashmore commented on src/customType.ts:
```diff
@@ -107,6 +108,18 @@ export const CustomTypeModelFieldType = {
 	Timestamp: "Timestamp",
 	UID: "UID",
+	/**
+	 * @deprecated - Legacy field type, use `Number` instead.
```
> Legacy field type. Use `Number` instead.

angeloashmore commented on src/customType.ts:
```diff
@@ -107,6 +108,18 @@ export const CustomTypeModelFieldType = {
+	 */
+	Separator: "Separator",
+	/**
+	 * @deprecated - Legacy field type, use `Slices` instead.
```
> Legacy field type. Use `Slices` instead.

angeloashmore commented on src/customType.ts:
```diff
@@ -201,10 +216,10 @@ export interface CustomTypeModelImageField<
+	config?: {
+		label?: string | null;
+		constraint?: CustomTypeModelImageConstraint;
+		thumbnails?: CustomTypeModelImageThumbnail<ThumbnailNames>[];
```
Correct me if I'm wrong, but the `readonly` keyword lets the type support `const` values.

```typescript
const type = { ... } as const
^ Arrays are now marked with readonly
```

If the `thumbnails` property is mutable (i.e. without `readonly`), `type` is no longer compatible with the `thumbnails` property.

When writing libraries, it might be a good practice for us to always use `readonly` for arrays unless we specifically need something to be mutable.

lihbr replied:
The issue here then is that internal types aren't readonly, and marking it as readonly here would make an internal type not being able to match the external type signature :thinking: not sure how we'd solve this.

lihbr replied:
Would `readonly thumbnail[] | thumbnail[]` work?

angeloashmore replied:
I tried adding `readonly` to all arrays in `customType.ts` and all the tests pass. I suppose it works because a mutable array is assignable to a readonly array type. The reverse is not true, however, which we will encounter when going from a Slice Machine plugin to SM internal code.

Does this mean we should be testing both directions since ultimately that's what we will be doing?

```typescript
/**
 * `@prismicio/types` extends `@prismicio/types-internal`
 */
expectType<prismicT.CustomTypeModelImageField>(
	{} as prismicTI.CustomTypes.Widgets.Nestable.Image,
);

/**
 * `@prismicio/types-internal` extends `@prismicio/types`
 */
expectType<prismicTI.CustomTypes.Widgets.Nestable.Image>(
	{} as prismicT.CustomTypeModelImageField,
);
```

We know the second test will always fail due to `types-internal`'s use of enums. After manually editing `types-internal`'s type to replace the enum with a string literal, the test passes, but only without `readonly`.

So I propose two things:

**1. Should we add `readonly` to `types-internal`?**

This will make `types-internal` more strict and easier to translate types between `types` and `types-internal`. This is at the expense of possibly causing trouble in other areas where arrays actually need to be mutable, such as constructing types programmatically.

**2. Should we replace the `types-internal`'s enum with a const object, like `types`?**

There should be no downside to this. It will, however, make it easier to translate types between `types` and `types-internal`. Using the type is more verbose, but a utility type could be made for better ergonomics.

arnaudlewis replied:
To answer on different aspect, from my knowledge, separator/Range are still supported and LegacySlices are still a thing on some repositories so I wouldn't just deprecate them. Range is not an alias, it has its own type so you would simply ignore all the Range that exists in users' content. If we want to make it an alias there is work to do on different front.
For the readonly, we could add them everywhere but we need to make sure that we don't have too many impacts considering that these models are used in API/CT API/prismic-mocks/editor/SM so I don't know what it would mean to change it but I'd be for it if it makes sense.
And for the WidgetTypes what would be the benefit to move WidgetTypes to a const object? I would have change `types` to an enum personally.
Anyway seems interesting and unifying those types could be good, even if some properties are useless I guess that's not a big problem and a config that is optional is for me very accurate, building a custom types on SM for instance will most probably end up in that situation. The goal for me would be to stay in sync with types-internal that actually reflects what's happening on our apps.

lihbr replied:
@angeloashmore 
> The reverse is not true, however, which we will encounter when going from a Slice Machine plugin to SM internal code.

Well, technically yes, in practice however since it'll pass through `io-ts` validator it should be fine(?)

---
@arnaudlewis
> separator/Range are still supported

We tried, `Range` is still supported and behaves as expected, so does `Separator`.

> LegacySlices are still a thing

Right, however the custom type editor crashes when trying to use them with `Choice` type, when using `Slices` they still work as expected. ([reference documentation from 2017](https://web.archive.org/web/20170523160728/https://prismic.io/docs/fields/slices))

> I wouldn't just deprecate them

Those (`Range`, `Separator`, legacy slices) are not documented anymore ([since at least 2019](https://web.archive.org/web/20190726004835/https://prismic.io/docs/javascript/getting-started/integrating-with-an-existing-javascript-project)) and buggy (writing room reacts weirdly to them, Slice Machine crashes or completely ignore them).

As these are external types, and that we don't expect users to be using them anymore (moving forward), we want to deprecate them on this library. Should some be reintroduced (realistically, I think only the `Range` slider might be) we'll be happy to update types again :)

> Range is not an alias

True, we separated them.

> you would simply ignore all the Range that exists in users' content.

@angeloashmore, maybe we can consider having `prismic-ts-codegen` taking them into account 🤔

> And for the WidgetTypes what would be the benefit to move WidgetTypes to a const object?

It's a bit tough to explain, but basically, if you were to type something after an enum (like a function argument) it would only accept values typed with that same enum. Unfortunately, this is not always the case/possible (when reading from a JSON for example, or when two different versions of a library are being used but the enum is the "same")

Here's a more concrete example:
![image](https://user-images.githubusercontent.com/25330882/176146950-17541549-3282-437a-9dcc-ebb7417aebc2.png)

And here's the full discussion we had on this: https://github.com/prismicio/prismic-types/issues/16

> Anyway seems interesting and unifying those types could be good, even if some properties are useless I guess that's not a big problem and a config that is optional is for me very accurate, building a custom types on SM for instance will most probably end up in that situation. The goal for me would be to stay in sync with types-internal that actually reflects what's happening on our apps.

🙌 we added type tests in that regard to ensure we'll stay in sync~

angeloashmore commented on src/customType.ts:
```diff
@@ -413,12 +436,22 @@ export interface CustomTypeModelTimestampField {
 }
 
+/**
+ * @deprecated - Legacy
```
I think we could use the same `@deprecated` description as the enum.

> Legacy field type. Do not use.

angeloashmore commented on src/customType.ts:
```diff
@@ -555,7 +599,14 @@ export interface SharedSliceModelVariation<
 }
+
+/**
+ * @deprecated - Legacy slice type, do not use.
```
> Legacy Slice type. Do not use.

angeloashmore reviewed (commented):
This looks great! As you mentioned, this is a breaking change, so we'll need to publish this at least as `v0.2.0`.

I left a few style and comment-related notes, but otherwise, it looks good to go.

lihbr commented:
Alright, fixed style and reverted back readonly's as discussed above.

Right now `@prismicio/types` are properly compatible with object types with `@prismicio/types-internal`, not the other way around.
We'll add test back for those when we get `@prismicio/types` object compatible with `@prismicio/types-internal`.

lihbr commented:
> Ongoing draft PR on types-internal to simplify the integration: [prismicio/prismic-types-internal#3](https://github.com/prismicio/prismic-types-internal/pull/3)

Added back two-ways test with [0b872c7](https://github.com/prismicio/prismic-types/pull/43/commits/0b872c7493d16044ec070f1d8975006035a4ebea), tested with https://github.com/prismicio/prismic-types-internal/pull/3 and everything passes cleanly 🙌

angeloashmore commented:
I corrected a small error in a test and made the link field tests more specific.

I'm going to merge and publish this PR. We can adjust the link field tests later if there's something we need to change.
